dra-evolution/pkg/api/capacity_types.go

```diff
@@ -1,170 +1,205 @@
-	Driver string `json:"driver,omitempty"`
-
-	// Attributes contains device attributes that are common to all devices
-	// in the pool.
-	// +optional
-	// +listType=atomic
-	Attributes []Attribute `json:"attributes,omitempty"`
-
-	// SharedResources are pooled resources that are shared by all devices
-	// in the pool. This is typically used when representing a
-	// partitionable device, and need not be populated otherwise.
+	// NodeName identifies the node which provides the devices
+	// if (and only if) they are local to a node. If this is empty,
+	// the devices exist outside of any specific node.
+	//
+	// NodeName and NodeSelector are mutually exclusive. If none is
+	// set, the devices are available in the entire cluster.
+	//
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain node name.
 	//
 	// +optional
-	// +listType=atomic
-	SharedResources []ResourceCapacity `json:"sharedResources,omitempty"`
+	NodeName string `json:"nodeName,omitempty" protobuf:"bytes,2,opt,name=nodeName"`
 
-	// Devices contains the individual devices in the pool.
+	// NodeSelector identifies all nodes that devices which are shared
+	// between nodes could be accessed from.
 	//
-	// +required
-	// +listType=atomic
-	Devices []Device `json:"devices,omitempty"`
-}
+	// NodeName and NodeSelector are mutually exclusive. If none is
+	// set, the devices are available in the entire cluster.
+	//
+	// +optional
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty"`
 
-// DevicePoolStatus contains the state of the pool as last reported by the
-// driver. Note that this will not include the allocations that have been made
-// by the scheduler but not yet seen by the driver. Thus, it is NOT sufficient
-// to make future scheduling decisions.
-type DevicePoolStatus struct {
-	DeviceStatuses []AllocatedDevice `json:"deviceStatuses,omitempty"`
-}
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain driver name.
+	DriverName string `json:"driverName" protobuf:"bytes,3,name=driverName"`
 
-// AllocatedDevice represents a device that has been allocated from the pool.
-type AllocatedDevice struct {
-	Name string
-	// Conditions contains the latest observation of the device's state.
-	Conditions []metav1.Condition `json:"conditions"`
+	ResourceModel `json:",inline" protobuf:"bytes,4,name=resourceModel"`
 
-	// ClaimUIDs contains the UIDs of the claims to which this device
-	// is allocated.
-	ClaimUIDs []types.UID
+	// To be decided: We could split ResourcePool into spec and status or
+	// extend NamedDevice below so that it reflects the current state of
+	// that device. Like the spec that is data that is coming from a driver,
+	// so adding more information is not a conceptual change. Could be
+	// added in any future release as an extension.
+	//
+	// Much harder is tracking "allocated for". That would be useful to
```

Comment:
Personally, I am leaning towards not using "spec". It just doesn't look like a normal user-created "spec" to me. "Data" then just becomes an unnecessary indirection. So simply top-level fields, as it is now?

I'd like to punt on adding a "status" for 1.31, at least in the initial KEP update. Like "consumable resources" that could be added later, both to the KEP and the implementation. We are already trying to do too much...

Reply:
What status do we need to track and at what granularity?  The lack of transactions scares me - I _guarantee_ we will fall out of sync on a regular basis, unless we make it absolutely bulletproof.  Can you imagine the FURY when a user has a cluster with (very expensive!!) GPUs, and then we just ... lose track of them?

Is it something like:
* Kubelet publishes the existence a device (DevicePool)
* Pod->Claim requests a device (DeviceClaim)
* Device deallocation controller puts a finalizer on the claim
* Scheduler matches them and updates the Pool.status
* Kubelet runs the pod
* User deletes the pod, cascades to the claim (who does this?)
* Device deallocation controller updates the Pool.status 
* Device deallocation controller removes the finalizer

What happens if there are multiple devices in different pools?
* Kubelet publishes the existence a device (DevicePool A)
* Kubelet publishes the existence a device (DevicePool B)
* Pod->Claim requests devices (DeviceClaim)
* Device deallocation controller puts a finalizer on the claim
* Scheduler matches them and updates the Pool A.status
* Scheduler gets a conflict on B.status - device was already taken!
* Scheduler crashes!
* Scheduler restarts

How does it recover?  I'm not seeing it.

Reply:
It absolutely is hard, no doubt. Let's stop thinking about how it might work and only assume that at some point we might have a status.

Would it be okay to have several top-level fields (i.e. no spec) and then later add a "status" field (doesn't even need to be a sub-resource, if we don't need that separation and it's not required by conventions)?

I think it would be okay.

Reply:
> Let's stop thinking about how it might work and only assume that at some point we might have a status

This puts all forward progress at risk, IMO.  If this requires elaborate, multi-stage choreography to be performed perfectly by shcedulers (which is understood to be an extension point),  that's bad.  Hand-waving "we'll fix it later" is worse.  We were finding bugs in the volume controller for years, and that's not even an extension point.

> Would it be okay to have several top-level fields (i.e. no spec) and then later add a "status" field 

I don't think we have any such resource, so I don't love the idea.

Reply:
> This puts all forward progress at risk, IMO.  If this requires elaborate, multi-stage choreography to be performed perfectly by schedulers (which is understood to be an extension point), that's bad.

I don't disagree. But we are not even making progress on the things where we could agree on some solution, so do we really think that we can pull off the design of yet another complex feature without putting everything at risk? :cry: 

>  I don't think we have any such resource, so I don't love the idea.

So let's use a "spec" field.

Reply:
> But we are not even making progress on the things where we could agree on some solution, so do we really think that we can pull off the design of yet another complex feature without putting everything at risk? 

If we have no idea how status works, everything is at risk.  We may decide that we went down a totally wrong path and the only way to make status work is something else, so all work on ratifying the rest is potentially wasted.

I'm not saying we have to exactly pin it down, but we have to have an approach in mind, don't we?  We are fundamentally pushing the bounds of what kube-apiserver can do sanely.

Reply:
I have an approach in mind. But we are already not making progress while trying to hash out "n" things in parallel. Making that "n + 1" is not going to help.

Reply:
How about this: we add a `DistributedAllocation` field to `ResourcePool`. If true, schedulers must use a future, yet to be defined "distributed allocation" protocol to use devices in the pool. If they don't support that, they must ignored the pool. If false, schedulers are allowed to use devices directly for claim allocation, without any coordination, as described now.

Future drivers can set that field based on their admin-controlled deployment configuration. As with the "privileged device" flag we can also allow a mutating webhook approach.

This fields removes the pressure to define the protocol now because we can role it out safely later.

Reply:
If we don't find a use for the field, we can remove it in the alpha -> beta transition.

Reply:
IMO, ResourcePool is 100% owned by the driver and no one else writes to it. That includes spec and status. That means that what gets written to status is just basic health information and an error code if there is unhealthiness. In fact it could be a sparsely populated status that is empty UNLESS the device is unhealthy.

This is completely separate from Tim's point about how we track the *allocations*. In 1.30 (please keep my honest, @pohly) these are captured in the claim status. When the scheduler starts up, it reads all the resources and all the allocated claims, and calculates and in-memory cache of the available resources. This is resilient to scheduler failures, and makes it pretty easy to deallocate and clean up.

This however, does not work if there are multiple (or multi-threaded) schedulers managing the same node. If we need to support multiple schedulers that schedule pods in parallel on the same set of resources, we could have some issues. If it were *just* node status we were writing to, then optimistic concurrency is sufficient to force retries of scheduling, since both schedulers would write to the node status and could identify that the status had changed since their last write. I **presume** this is how multiple schedulers avoid over committing node resources in Kubernetes today (ie, even with just standard resources).

Given though that we also now have schedulers writing to *independent claim objects*, we have to be more careful. We need to write the status to each of the claim objects, and then write the node status *last*. If two schedulers scheduler pods to the same node, the second scheduler would see a "this resource has changed" message. At that point, it needs to back out *all of the scheduling choices it made*. If we do that, I think we are safe...

Reply:
I had been assuming that status was on device pool.  Putting it on claim is maybe not as bad.  I'm mostly worried about the failure modes.  If the worst case is that 2 pods land on a node and the node says "no, that device is in use", I guess it's not worse than what we do with CPU and memory.  But I want to be convinced that is REALLY the worst case.

But if I have a pod with multiple claims, and the scheduler crashes before finishing, I guess we have enough information to finish?  Claim A is resolved with `{node, driver, device}`, so that can either be backed out or continued.  I think?

Your point about different claim objects is something I had not considered, but I think is no different than any case of "pod killed"?

> write the node status last

Who is writing node status?  We don't do that for regular resources.

Reply:
> But if I have a pod with multiple claims, and the scheduler crashes before finishing, I guess we have enough information to finish? Claim A is resolved with `{node, driver, device}`, so that can either be backed out or continued. I think?

It should be backed out. If we were using the node status revision to manage this (which it sounds like we will not, based on the below), we would still need to back it out. The reason is that the status revision may have advanced since we made the decision, and since we are not holding the old status revision and sending that for optimistic concurrency, we could get an overcommit if we don't back it out. However, this is a moot point given the info below.

> Who is writing node status? We don't do that for regular resources.

Ah. That's why I said "I presume", and I was wrong. So, we already could overcommit resources. This wouldn't be any different. But we would have to make sure that the driver authors know that this can happen, otherwise I can see a naive driver mounting the device in the two different, unrelated containers.

Reply:
> I presume this is how multiple schedulers avoid over committing node resources in Kubernetes today (ie, even with just standard resources).

They don't. You end up with over-committing if multiple different schedulers schedule to the same node.

> I had been assuming that status was on device pool. Putting it on claim is maybe not as bad.

But then we don't get coordination. When there's contention for devices, we end up with different pods being scheduled although not all can run - exactly the issue we try to solve.

> If the worst case is that 2 pods land on a node and the node says "no, that device is in use", I guess it's not worse than what we do with CPU and memory. But I want to be convinced that is REALLY the worst case.

It is worse for network-attached devices. Then the node will not know that the device is in use unless it constantly watches all claims to verify that there is no over-commit - that itself is then racy.

To me, extending the ResourcePool status as the authoritative place for "device is reserved" makes the most sense. It leads to various error scenarios that we need to identify and handle, but I think it's doable. I know it's tempting and a much more interesting problems than the API design topics,  but let me ask again: do we really, absolutely need to design this now?

I thought we wanted to focus on node-local resources where the traditional behavior of the scheduler ("I own the resources on this node") is fine.

Reply:
If we have multiple schedulers making requests from multiple pools, and we store allocations in the pools, we have to deal with finding that a claim cannot be satisfied at write time:
- scheduler one chooses pool X, Y
- scheduler two chooses pool X, Y
- scheduler one allocates X
- scheduler two allocates Y
- scheduler one tries to allocate Y but gets hit with "resource version has changed"
- scheduler two tries to allocate X but gets hit with "resource version has changed"

At this point I think both schedulers will need to back out all scheduling choices and try again. Similarly if one crashes.

I think this should be OK. We could have an extended battle but theoretically someone should win the race. Unless there are like 5 pools, in which case it could get messy. Maybe some back off with randomness would help.

Reply:
> It is worse for network-attached devices. Then the node will not know that the device is in use unless it constantly watches all claims to verify that there is no over-commit - that itself is then racy.

All the more reason for those to be in their own pool type...

> I think this should be OK. We could have an extended battle but theoretically someone should win the race. Unless there are like 5 pools, in which case it could get messy. Maybe some back off with randomness would help.

This is a *little* worse as I think more about it. The driver also writes to this object, so it can trigger the "resource version has changed" as well.

We also will need to carefully think through how we back out the changes. In the case of a "resource version has changed", we are in the midst of scheduling, so we should have in-memory the right context to go back and back out all the changes we made and restart the scheduling.

In the case of a crash, we will start up and find a partially scheduled pod. We will need to backtrack from the pools to the claims, to the pod. Which is ugly but do-able.

We should think through how the finalizers get added and make sure we're not creating some deadlock or other issues. I think it should be OK, I just haven't thought through the detail.

An alternative that would eliminate the concurrent modification issue (but not the crash recovery one) would be to use a per-node mutex whenever we write to a pool belonging to a node. This would need to be via a Lease or something. The performance overhead is *probably* not worth it. But it would simplify a lot of things and likely avoid some bugs or unexpected interactions.

In generally, I think we would make our lives a lot easier if we kept to single writers for objects. So: driver owns and is only writer to pools; scheduler owns and is single writer to claims (though they are generated by a separate controller).

Reply:
> All the more reason for those to be in their own pool type... 

Sorry, I don't see how that helps. We want reservation to work also for node-local devices, because then a DRA controller which does classic DRA can co-exist with a normal scheduler, or multiple schedulers can target the same node (not recommended because of RAM and CPU, but perhaps that's okay). So all pool types would have to support the same extra fields for reservation.

> In generally, I think we would make our lives a lot easier if we kept to single writers for objects. So: driver owns and is only writer to pools; scheduler owns and is single writer to claims (though they are generated by a separate controller).

I don't see how that helps with coordination. On which object do two schedulers coordinate which want to allocate the same device to two different claims?

I agree that distributed updating of ResourcePool and ResourceClaim objects in a certain order will be possible. I could call out various corner cases and how I would handle them, but I am going to resist the temptation. That is not a good use of my and your time. We need an entire design document for that, not some scattered GitHub PR comments.

Reply:
> I don't see how that helps with coordination. On which object do two schedulers coordinate which want to allocate the same device to two different claims?

They don't, it would have the same weakness when using multiple schedulers as we do for CPU and memory. But boy would the code be simpler.

> We need an entire design document for that, not some scattered GitHub PR comments.

Agree, this is PR has become unwieldy. May I suggest you pick the "currently agreed" things, resolve them, and anything "under discussion" you add to an "open questions" document, and we merge this? That't the strategy I used in the k8srm-prototype. It's not ideal because we have to actually *come back* to the open questions. But keeping track in this PR is gotten very difficult.